File: PrefixSum/3381. Maximum Subarray Sum With Length Divisible by K/solution.py
class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        presum = list(accumulate(nums, initial=0))

        res = -inf
        seen = defaultdict(lambda: inf)
        for size in range(1, n + 1):
            if size % k == 0:
                res = max(res, presum[size])

            if (size % k) in seen:
                res = max(res, presum[size] - seen[size % k])

            seen[size % k] = min(seen[size % k], presum[size])
        return res

File: PrefixSum/3381. Maximum Subarray Sum With Length Divisible by K/solution.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    long long maxSubarraySum(vector<int>& nums, int k) {
        unordered_map<int,long long> m = {{0,0}};
        long long res = LLONG_MIN, presum = 0;
        for (int i=0; i<nums.size(); i++) {
            presum += nums[i];
            int key = (i+1)%k;

            if (m.find(key) != m.end()) {
                res = max(res, presum - m[key]);
            }

            if (m.find(key) == m.end()) {
                m[key] = presum;
            } else {
                m[key] = min(m[key], presum);
            }
        }
        return res;
    }
};